apollo-federation/src/sources/connect/models/validation.rs

```diff
@@ -49,3 +39,225 @@
+    source_names
+        .into_iter()
+        .counts()
+        .into_iter()
+        .filter_map(|(name, count)| {
+            (count > 1).then_some(ValidationError::DuplicateSourceName(name))
+        })
+        .chain(errors)
+        .collect()
+}
+
+fn validate_source(directive: &Component<Directive>) -> Result<SourceDirective, ValidationError> {
+    let name = SourceName::try_from(directive)?;
+    let url_error = directive
+        .arguments
+        .iter()
+        .find(|arg| arg.name == "http")
+        .and_then(|arg| {
+            let value = arg
+                .value
+                .as_object()?
+                .iter()
+                .find_map(|(name, value)| (name == "baseURL").then_some(value))?
+                .as_str()?;
+            let url = match Url::parse(value) {
+                Err(inner) => {
+                    return Some(ValidationError::SourceUrl {
+                        inner,
+                        source_name: name.clone(),
+                    })
+                }
+                Ok(url) => url,
+            };
+            if url.scheme() != "http" && url.scheme() != "https" {
+                return Some(ValidationError::SourceScheme {
+                    scheme: url.scheme().to_string(),
+                    source_name: name.clone(),
+                });
+            }
+            None
+        });
+    let mut errors = Vec::new();
+    if matches!(name, SourceName::Invalid(_)) {
+        errors.push(ValidationError::InvalidSourceName(name.clone()));
+    }
+    if let Some(url_error) = url_error {
+        errors.push(url_error);
+    }
+    Ok(SourceDirective { name, errors })
+}
+
+/// A `@source` directive along with any errors related to it.
+struct SourceDirective {
+    name: SourceName,
+    errors: Vec<ValidationError>,
+}
+
+/// The `name` argument of a `@source` directive.
+#[derive(Clone, Debug)]
+pub enum SourceName {
+    /// A perfectly reasonable source name.
+    Valid(String),
+    /// Contains invalid characters, so it will have to be renamed. This means certain checks
+    /// (like uniqueness) should be skipped. However, we have _a_ name, so _other_ checks on the
+    /// `@source` directive can continue.
+    Invalid(String),
+}
+
+impl From<SourceName> for String {
+    fn from(source_name: SourceName) -> String {
+        match source_name {
+            SourceName::Valid(name) | SourceName::Invalid(name) => name,
+        }
+    }
+}
+
+impl TryFrom<&Component<Directive>> for SourceName {
+    type Error = ValidationError;
+
+    fn try_from(directive: &Component<Directive>) -> Result<Self, Self::Error> {
+        let str_value = directive
+            .arguments
+            .iter()
+            .find(|arg| arg.name == "name")
+            .ok_or(ValidationError::MissingSourceName)?
+            .value
+            .as_str()
+            .ok_or(ValidationError::SourceNameType)?;
+        if str_value.chars().all(|c| c.is_alphanumeric() || c == '_') {
```

Comment:
Does this mean that an empty string is allowed as a name?

Reply:
It's not! Nice catch, will add in that check as well.